packages/compass-aggregations/src/utils/pipeline-storage.ts

```diff
@@ -57,88 +41,84 @@ function savedPipelineToText(pipeline: StoredPipeline['pipeline']): string {
   }
 }
 
-function hasAllRequiredKeys(pipeline?: any): pipeline is StoredPipeline {
-  return (
-    pipeline &&
-    typeof pipeline === 'object' &&
-    ['id', 'name', 'namespace'].every((key) => key in pipeline)
-  );
-}
+const PipelineSchema = z
+  .object({
+    id: z.string(),
+    name: z.string(),
+    namespace: z.string(),
+    comments: z.boolean().optional(),
+    autoPreview: z.boolean().optional(),
+    collationString: z.string().optional(),
+    pipeline: z
+      .array(
+        z.object({
+          stageOperator: z.string(),
+          isEnabled: z.boolean(),
+          stage: z.string(),
+        })
+      )
+      .optional()
+      .describe('Legacy property to stored pipeline. Use pipelineText.'),
+    host: z.string().nullable().optional(),
+    pipelineText: z.string().optional(),
+    lastModified: z
+      .number()
+      .transform((x) => new Date(x))
+      .optional(),
+  })
+  .transform((input) => {
+    const {
+      pipeline: legacyPipelineArray,
+      pipelineText,
+      ...restOfInput
+    } = input;
+    return {
+      ...restOfInput,
+      pipelineText: pipelineText ?? savedPipelineToText(legacyPipelineArray),
+    };
+  });
+
+export type StoredPipeline = z.output<typeof PipelineSchema>;
+
+type SaveablePipeline = Omit<z.input<typeof PipelineSchema>, 'pipeline'>;
 
 export class PipelineStorage {
+  private readonly userData: UserData<typeof PipelineSchema>;
+  constructor(basePath?: string) {
+    this.userData = new UserData(PipelineSchema, {
+      subdir: 'SavedPipelines',
+      basePath,
+    });
+  }
   async loadAll(): Promise<StoredPipeline[]> {
     try {
-      const dir = getDirectory();
-      const files = (await fs.readdir(dir))
-        .filter((file) => file.endsWith('.json'))
-        .map((file) => path.join(dir, file));
-
-      return (
-        await Promise.all(files.map((filePath) => this._loadOne(filePath)))
-      ).filter(Boolean) as StoredPipeline[];
+      const { data } = await this.userData.readAll({
+        ignoreErrors: false,
+        readFileStats: true,
+        mergeStats(input, stats) {
+          return {
+            ...input,
+            lastModified: Number(stats.mtimeMs),
+          };
+        },
```

Comment:
Stats seem to be merged for load all but not for load one, is this intentional change?

Reply:
nice catch! missed it completely.